borgbase/vorta
# Wait on network to be available before trying to run backups against remote repos

### Description

When the scheduler tries to start a job on a remote repo remote and there is definitely no network connection
defer the job.
Also, check for said deferred jobs on an interval. If any exist, check the current status of the network and if it is up then start all of these backups.

### Related Issue
https://github.com/borgbase/vorta/issues/2133

### Motivation and Context
Running a backup against a remote repo will fail without a network connection. 
This results in a notification on default settings, stress (because why are my backups failing) and could maybe result in backups not happening.
On my Linux systems configured to sleep, the first backup after resuming always seems to fail because the network has not finished activating yet.
This should allow that backup to just wait for a little while.

As a side benefit, this should also defer backups that happen in the normal course when there is no possible way they could work (i.e. when in airplane mode).

I thought about these changes for a little and its what I came up with. However, if the maintainers have a different approach they would like to try I'm more than willing to rework my PR or open a new one with that approach. I do also have a mac that I can in theory develop the stubbed mac implementation here, but I'm less familiar with how to programatically access the state of the mac network. If this approach is good, I can flesh that out also.

### How Has This Been Tested?

1. Activate airplane mode from kde plasma network manager
1. Manipulate the event log in settings.db so that the scheduled backup will fire immediately
4. Start vorta
5. Observe the new logs indicating that the scheduled task is being postponed because no network
6. Re-enable the network
7. Observe the backup completing.
8. Observe that network polling stops

I had a little bit of trouble figuring out how to manipulate vorta's database state so that I could actually start vorta, then put my system to sleep. Vorta always wanted to backup immediately. I assume this should be a reasonable facsimile of the behavior. 

Regarding actual tests, I am not an experienced python developer and running nox locally seems to fail to collect tests in test_darwin.py which I wouldn't expect to execute at all on my system. That said, I don't think I changed anything covered by unit tests.

### Screenshots (if appropriate):

### Types of changes


### Checklist:


- [x] I have read the [CONTRIBUTING](https://vorta.borgbase.com/contributing/) guide.
- [ ] All new and existing tests passed. * see note above re nox.


*I provide my contribution under the terms of the [license](./../LICENSE.txt) of this repository and I affirm the [Developer Certificate of Origin][dco].*

[dco]: https://developercertificate.org/
